pkg/frontend/openshiftcluster_putorpatch_test.go

```diff
@@ -0,0 +1,396 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"net/http"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/v20191231preview"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/env"
+	mock_database "github.com/Azure/ARO-RP/pkg/util/mocks/database"
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+	"github.com/Azure/ARO-RP/test/util/listener"
+	"github.com/Azure/ARO-RP/test/util/matcher"
+)
+
+type dummyOpenShiftClusterValidator struct{}
+
+func (*dummyOpenShiftClusterValidator) Static(interface{}, *api.OpenShiftCluster) error {
+	return nil
+}
+
+func (*dummyOpenShiftClusterValidator) Dynamic(context.Context, *api.OpenShiftCluster) error {
+	return nil
+}
+
+func expectAsyncOperationDocumentCreate(asyncOperations *mock_database.MockAsyncOperations, key string, provisioningState api.ProvisioningState) {
+	asyncOperations.EXPECT().
+		Create((*matcher.AsyncOperationDocument)(
+			&api.AsyncOperationDocument{
+				OpenShiftClusterKey: key,
+				AsyncOperation: &api.AsyncOperation{
+					InitialProvisioningState: provisioningState,
+					ProvisioningState:        provisioningState,
+				},
+			}),
+		)
+}
+
+func TestPutOpenShiftCluster(t *testing.T) {
+	ctx := context.Background()
+
+	apis := map[string]*api.Version{
+		"2019-12-31-preview": api.APIs["2019-12-31-preview"],
+	}
+	apis["2019-12-31-preview"].OpenShiftClusterValidator = func(env.Interface, string) api.OpenShiftClusterValidator {
+		return &dummyOpenShiftClusterValidator{}
+	}
```

Comment:
If I follow pointers and how we initialise versions correctly, this bit will be basically the same as doing this:

```go
api.APIs["2019-12-31-preview"].OpenShiftClusterValidator = func(env.Interface, string) api.OpenShiftClusterValidator {
	return &dummyOpenShiftClusterValidator{}
}
```

And I think it will introduce side effects for other test functions that will run after `TestPutOpenShiftCluster` and rely on `api.APIs["2019-12-31-preview"].OpenShiftClusterValidator`.

That is one of the reasons I had the `Version` interface and got rid of global `api.APIs` and replaced it with explicit version instantiation.